include/xtensor/xstrides.hpp

```diff
@@ -142,6 +142,123 @@ namespace xt
+        normal = 1, ///< Normal stride corresponding to storage.
+        bytes = 2, ///< Normal stride in bytes.
+    };
+
+    /**
+    * @ingroup strides
+    * @brief strides
+    *
+    * Get strides of an object.
+    * @param a an array
+    * @return array
+    */
+    template <class E>
+    inline auto strides(const E& e, stride_type type = stride_type::normal) noexcept
+    {
+        using strides_type = typename E::strides_type;
+        using ret_type = typename strides_type::value_type;
+        strides_type ret = e.strides();
+        auto shape = e.shape();
+
+        if (type == stride_type::internal) {
+            return ret;
+        }
+
+        for (std::size_t i = 0; i < ret.size(); ++i) {
+            if (shape[i] == 1) {
+                if (e.layout() == layout_type::row_major) {
+                    ret[i] = std::accumulate(
+                        shape.cbegin() + i + 1,
+                        shape.cend(),
+                        static_cast<ret_type>(1),
+                        std::multiplies<ret_type>()
+                    );
+                }
+                else
+                {
+                    ret[i] = std::accumulate(
+                        shape.cbegin(),
+                        shape.cbegin() + i,
+                        static_cast<ret_type>(1),
+                        std::multiplies<ret_type>()
+                    );
+                }
+            }
+        }
+
+        if (type == stride_type::bytes) {
+            ret_type f = static_cast<ret_type>(sizeof(typename E::value_type));
+            std::for_each(ret.begin(), ret.end(), [f](auto& c){ c *= f; });
+        }
+
+        return ret;
+    }
+
+    /**
+    * @ingroup strides
+    * @brief strides
+    *
+    * Get stride of an object along an axis.
+    * @param a an array
+    * @return integer
+    */
+    template <class E>
+    inline auto strides(const E& e, std::size_t axis, stride_type type = stride_type::normal) noexcept
+    {
+        using strides_type = typename E::strides_type;
+        using ret_type = typename strides_type::value_type;
+
+        ret_type ret = e.strides()[axis];
+
+        if (type == stride_type::internal) {
+            return ret;
+        }
+
+        if (ret == 0) {
+            if (e.shape(axis) == 1) {
+                if (e.layout() == layout_type::row_major) {
+                    ret = std::accumulate(
+                        e.shape().cbegin() + axis + 1,
+                        e.shape().cend(),
+                        static_cast<ret_type>(1),
+                        std::multiplies<ret_type>()
+                    );
+                }
+                else
+                {
+                    ret = std::accumulate(
+                        e.shape().cbegin(),
+                        e.shape().cbegin() + axis,
+                        static_cast<ret_type>(1),
+                        std::multiplies<ret_type>()
+                    );
+                }
```

Comment:
I think it would make sense to have an "impementation" function to avoid code duplication in the different overloads.

Reply:
Alright! It think it should be good to go like this. I realised that I did not account for `xt::layout_type::dynamic`. BTW, I think there is not static way to specialise for the `layout_type` is there?

Reply:
OK. In that case I think we can merge.